### PR TITLE
[FEAT] UserAnswer Create API 구현 

### PR DESCRIPTION
**Pull request by 5hseok**
## #️⃣ Related Issue

Linked Feature backlog:  #53 

## 📝 Purpose of PR

온보딩 시 사용자가 입력한 데이터를 저장한다. 

## Contents

온보딩 시 사용자가 입력한 데이터를 받아 저장한다.
만약, 기존에 저장한 내역이 있다면 새로 들어온 데이터로 UserAnswer 테이블의 내용을 교체한다. 

## Checklist

-   [ ] Does commit messages follow the convention?
-   [ ] Are variable names brief but descriptive?
-   [ ] Is the code clean and easy to understand?
-   [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
-   [ ] Does the test successfully test edge and corner cases?

## 💬 Review Requirements (Optional)

Is there anything you can praise about this PR? Start with praise.
Make a review, leaving kind comments and suggesting changes where needed (to resolve the above).

### CODE REVIEW

File: src/main/java/ac/dnd/dodal/ui/user/request/CreateUserAnswerRequestDto.java
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/ui/user/request/CreateUserAnswerRequestDto.java:1]: question과 answer이 id 없이 string으로만 전달된 상황이라 id값을 받아오는 게 불가능할 것 같아요
앞선 pr에도 있지만 id말고 string으로 저장하는 건 어떨까요?

File: src/main/java/ac/dnd/dodal/domain/onboarding/exception/OnBoardingExceptionCode.java
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/domain/onboarding/exception/OnBoardingExceptionCode.java:1]: onboarding exception이면 ANS보다 ONB가 더 적절하지 않을까요?
ANS면 계획 피드백하는 것도 엮일 수가 있을 것 같아요!

File: src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java [hunk 1]
```diff
@@ -30,4 +49,53 @@ public User createUserBySocialSignUp(OAuthUserInfoRequestDto authSignUpRequestDt
...
+    public void createUserAnswer(Long userId, CreateUserAnswerRequestDto createUserAnswerRequestDtoList) {
+        // 들어온 데이터의 수가 전체 질문의 수와 일치하는 지 확인
+        // 온보딩이 도중에 종료되었을 경우, 다시 처음부터 시작할 수 있도록 유도
+        if (createUserAnswerRequestDtoList.data().size() != questionRepository.countAllQuestions()) {
+            throw new OnBoardingBadRequestException(OnBoardingExceptionCode.INVALID_ANSWER_COUNT,
+                    "The number of answers does not match the total number of questions.");
+        }
+
+        // 사용자 조회
+        User user = userCommandRepository.findById(userId)
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:63]: 이부분은 `@UserId`에서 검증되는 게 아닌가요?

File: src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java [hunk 2]
```diff
@@ -30,4 +49,53 @@ public User createUserBySocialSignUp(OAuthUserInfoRequestDto authSignUpRequestDt
...
+                .orElseThrow(() -> new UserNotFoundException(UserExceptionCode.NOT_FOUND_USER));
+
+        // 사용자의 답변이 존재하는지 확인
+        List<UserAnswer> existUserAnswer = userAnswerRepository.findAllByUserId(user);
+        if (existUserAnswer != null && !existUserAnswer.isEmpty()) {
+            log.info("existUserAnswer: {}", existUserAnswer);
+            userAnswerRepository.deleteAll();
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:70]: 사용자의 답변이 이미 존재한다면 예외를 던지는 게 맞지 않을까요?
아니면 온보딩 데이터를 전달할 때 해당 유저가 온보딩이 필요한지 유무를 판단하여 거기서 예외처리를 할 수도 있을 것 같아요

File: src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java [hunk 3]
```diff
@@ -30,4 +49,53 @@ public User createUserBySocialSignUp(OAuthUserInfoRequestDto authSignUpRequestDt
...
+        List<Long> questionIds = createUserAnswerRequestDtoList.data().stream()
+                .map(CreateUserAnswerRequestDto.UserAnswerData::questionId)
+                .distinct()
+                .collect(Collectors.toList());
+
+        // N+1을 피하기 위해 해당 questionId들을 가진 모든 Answer들을 미리 한 번에 조회
+        List<Answer> answers = answerRepository.findAllByQuestionId(questionIds);
+
+        // (questionId, answerId) 조합을 키로 하는 Map 생성
+        Map<String, Answer> answerMap = answers.stream()
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:83]: 이것을 map으로 생성해야하는 이유가 있을까요?
받아올 때 순서나 question을 묶어서 한번에 받아오는 게 더 적절하지 않을까요?

File: src/main/java/ac/dnd/dodal/application/onboarding/repository/QuestionRepository.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+import ac.dnd.dodal.domain.onboarding.model.Question;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface QuestionRepository extends JpaRepository<Question, Long> {
+
+    @Query("SELECT COUNT(q) FROM Question q")
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/application/onboarding/repository/QuestionRepository.java:11]: deletedAt is NULL 추가해야할 것 같습니당

File: src/main/java/ac/dnd/dodal/application/onboarding/repository/AnswerRepository.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+import java.util.Optional;
+
+@Repository
+public interface AnswerRepository extends JpaRepository<Answer, Long> {
+
+    @Query("SELECT a FROM Answer a WHERE a.question.id IN :questionIds")
+    List<Answer> findAllByQuestionId(
```
Comment (sueelly) [inline src/main/java/ac/dnd/dodal/application/onboarding/repository/AnswerRepository.java:16]: deletedAt is NULL 추가해야할 것 같습니다
아니면 deleted에 대해 따로 어플리케이션 레벨에서 처리하는 로직을 추가하는 방식으로요!

**Review by sueelly — CHANGES_REQUESTED** [review]

File: src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java (hunk 2 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:70]: 온보딩 데이터가 대거 수정되어 다시 온보딩 처리를 해준다고 했을 때, 예외를 던지는 것보다는 사용자가 가지는 온보딩 데이터를 교체하는 것이 더 적절하다고 생각했습니다! 

예외처리를 하게 된다면 로그인 시 사용자가 이미 온보딩을 진행했다면 온보딩 과정을 처리하지 않는다는 응답을 로그인 응답에서 보내주면 될 것 같은데, 이렇게 되면 사용자는 온보딩을 진행한 이후로 다시는 온보딩을 하지 못하게 되는 것이라고 생각해서 확장성이나 케이스를 생각해봤을 때 다시 온보딩 설정을 하고 싶을 때 다시 온보딩을 진행할 수 있는 것이 더 자연스럽다고 생각했습니다. 

저희가 아직은 온보딩 데이터를 바꾸는 플로우나 기능은 없지만요

File: src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java (hunk 3 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:83]: 한번에 받아온다는 게 무슨 의미일까요?

Map을 사용한 이유는 그룹핑이나 정렬을 하지 않고 탐색 시 시간복잡도를 줄이기 위해 사용했습니다.

File: src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java (hunk 1 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:63]: User 검증은 되었는데, userAnswerRepository에서 user 객체를 다루기 때문에 사용했습니다. 

User를 안 가져오고 userId로 쓰려면 

`    @Query("SELECT ua FROM UserAnswer ua WHERE ua.user.id = :userId")
`
이런 식으로 쿼리를 써야하거나 UserAnswer 엔티티 생성자에서 userId를 사용해서 User 객체로 변환하는 과정이 추가되는 것 같아서요!

Comment (sueelly) [reply src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:63]: 오 그러네요 생성할 때는 User가 필요하군요

File: src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java (hunk 2 above)
Comment (sueelly) [reply src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:70]: 온보딩이라는 게 처음에 가이드해주는 역할같은 거라, 다시 할 일은 없지 않을까 싶어요. 물론 아직 논의된 바는 없지만요.
그리고 이러한 확정성을 고려하는 경우에, 첫 create와 update하는 것을 구분하여 구현하는 게 어떨까요?

그리고 저기서 어떤 식별자 없이 deleteAll을 해버리면 모든 사용자의 데이터가 삭제될 것 같습니다!

File: src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java (hunk 3 above)
Comment (sueelly) [reply src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:83]: client에서 answer만 받아오는 게 아니라 question과 order도 한번에 받아오는 것을 뜻합니당
만에 하나 같은 answer이 존재한다하여도 식별 가능할 뿐더러 매번 모든 조합의 키를 생성하여 그 중 하나를 찾아내는 게 번거롭지 않나 싶어요

File: src/main/java/ac/dnd/dodal/ui/user/request/CreateUserAnswerRequestDto.java
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/ui/user/request/CreateUserAnswerRequestDto.java:1]: 아 제가 앞선 PR에서 커밋 정리하다가 내용이 하나 날아간 것 같네요.
```
{
    "code": "COM000",
    "message": "Success",
    "data": [
        {
            "questionId": 1,
            "questionContent": "현재 관심 있는 목표 분야를 하나만 선택해 주세요.",
            "subContent": "분야별로 최적의 인사이트를 추천해드릴게요!",
            "order": 1,
            "answers": [
                {
                    "answerId": 1,
                    "content": "건강/운동"
                },
                {
                    "answerId": 2,
                    "content": "학습/자기계발"
                },
                {
                    "answerId": 3,
                    "content": "커리어/직업"
                },
                {
                    "answerId": 4,
                    "content": "취미/여가"
                }
            ]
        },
        {
            "questionId": 2,
            "questionContent": "계획을 세울 때 선호하는 방식을 모두 골라주세요.",
            "subContent": "당신의 스타일에 딱 맞는 실행 전략을 찾아드릴게요!",
            "order": 2,
            "answers": [
                {
                    "answerId": 1,
                    "content": "구체적인 단위별 계획"
                },
                {
                    "answerId": 2,
                    "content": "큰 방향성만 잡기"
                },
                {
                    "answerId": 3,
                    "content": "유연하게 접근하기"
                },
                {
                    "answerId": 4,
                    "content": "체크리스트 활용"
                }
            ]
        },
        {
            "questionId": 3,
            "questionContent": "그동안 계획 설정 시 어려웠던 점을 모두 선택해주세요.",
            "subContent": "당신의 고민에 맞춘 계획 개선 팁을 제공해드릴 수 있어요!",
            "order": 3,
            "answers": [
                {
                    "answerId": 1,
                    "content": "목표 구체화"
                },
                {
                    "answerId": 2,
                    "content": "시간 관리"
                },
                {
                    "answerId": 3,
                    "content": "우선순위 결정"
                },
                {
                    "answerId": 4,
                    "content": "실행 동기 부족"
                }
            ]
        }
    ]
}
```
이런 식으로 데이터를 내려줄 때 id를 포함해서 내려줄 수 있습니다!

File: src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java (hunk 2 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:70]: 아 그러네요 허허 수정해서 다시 올려보겠습니다! 
일단은 이미 데이터가 존재한다면 예외를 던지는 식으로 수정할게요!

File: src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java (hunk 3 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:82]: 우선 client에서는 questionId와 answerId를 같이 받고있긴 합니다. 
```
// 사용자 답변 저장
        for (CreateUserAnswerRequestDto.UserAnswerData dto : createUserAnswerRequestDtoList.data()) {
            Answer answer = answerRepository.findByQuestionIdAndAnswerId(dto.questionId(), dto.answerId())
                            .orElseThrow(()-> new OnBoardingNotFoundException(OnBoardingExceptionCode.NOT_FOUND_ANSWER));

            userAnswerRepository.save(new UserAnswer(answer, user));
        }
```
이러면 questionId랑 answerId가 묶인 상태로 질문 개수만큼 생기고 이 개수만큼 쿼리를 보내야 하니까 N+1 문제가 생기지 않나요? 

questionId랑 answerId가 묶여있는 리스트로 데이터를 가져오려고 하니까 JPA나 JPQL에서는 지원하지 않아 Native query를 사용해야 하는 것 같아요.

근데 Native Query를 사용하는 것보다는 이렇게 하는 게 낫지 않을까요? 시간복잡도 자체는 O(n)으로 나올거에요.

File: src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java (hunk 2 above)
Comment (5hseok) [reply src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java]: 수정해두었습니다!

File: src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java (hunk 3 above)
Comment (sueelly) [reply src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:82]: 프론트에서 백에서 넘겨준 값을 언제나 사용한다면 이것이 맞는 값인지에 대한 검증을 하지 않아도될 것 같아서요!

Comment (sueelly) [reply src/main/java/ac/dnd/dodal/application/user/service/UserCommandService.java:82]: 우선 해당 로직으로도 정상 작동할 것 같아서, 추후 리팩토링한다면 다시 논의하시죠!
